File: quant-ia/core/market_data_quality.py
from __future__ import annotations

import math
import time
from collections import Counter, deque
from dataclasses import dataclass
from typing import Any

TIMEFRAME_SECONDS = {
    "1m": 60,
    "3m": 180,
    "5m": 300,
    "15m": 900,
    "30m": 1800,
    "1h": 3600,
    "4h": 14400,
    "1d": 86400,
}

_metrics: Counter[str] = Counter()
_incidents: deque[dict[str, Any]] = deque(maxlen=200)
_event_claims: dict[str, float] = {}


def normalize_symbol(symbol: str) -> str:
    value = str(symbol).strip().upper().replace("_", "-").replace("/", "-")
    if "-" in value:
        return value
    for quote in ("USDT", "USDC", "USD"):
        if value.endswith(quote):
            base = value[: -len(quote)]
            return f"{base}-{'USDT' if quote == 'USD' else quote}"
    return value


def normalize_timeframe(timeframe: str) -> str:
    value = str(timeframe).strip().lower()
    if value not in TIMEFRAME_SECONDS:
        raise ValueError(f"Unsupported timeframe: {timeframe}")
    return value


def canonical_market_event_id(symbol: str, timeframe: str, open_time_ms: int) -> str:
    value = int(open_time_ms)
    if value <= 0:
        raise ValueError(f"Invalid candle open timestamp: {open_time_ms}")
    return f"md:v1:bingx:{normalize_symbol(symbol)}:{normalize_timeframe(timeframe)}:{value}"


def _incident(kind: str, symbol: str, timeframe: str, detail: str, now: float) -> None:
    _metrics[kind.lower()] += 1
    _incidents.append({
        "type": kind,
        "symbol": normalize_symbol(symbol),
        "timeframe": timeframe,
        "occurredAt": now,
        "detail": detail,
    })


def _float(item: dict, *keys: str) -> float:
    for key in keys:
        if key in item:
            return float(item[key])
    return math.nan


def _open_time_ms(item: dict) -> int:
    for key in ("time", "openTime", "open_time", "timestamp"):
        if key in item:
            value = float(item[key])
            return int(value * 1000 if value < 10_000_000_000 else value)
    return 0


def sanitize_candles(
    symbol: str,
    timeframe: str,
    rows: list[dict],
    now_ms: int | None = None,
    stale_after_intervals: int = 2,
) -> dict[str, Any]:
    tf = normalize_timeframe(timeframe)
    interval_ms = TIMEFRAME_SECONDS[tf] * 1000
    now_ms = int(time.time() * 1000) if now_ms is None else int(now_ms)
    out_of_order = sum(
        1 for left, right in zip(rows, rows[1:])
        if _open_time_ms(right) < _open_time_ms(left)
    )
    unique: dict[int, dict] = {}
    duplicates = 0
    invalid = 0
    for row in rows:
        try:
            open_time = _open_time_ms(row)
            open_price = _float(row, "open", "o")
            high = _float(row, "high", "h")
            low = _float(row, "low", "l")
            close = _float(row, "close", "c")
            volume = _float(row, "volume", "vol", "v")
            values = (open_price, high, low, close, volume)
            valid = (
                open_time > 0
                and open_time % interval_ms == 0
                and all(math.isfinite(value) for value in values)
                and min(open_price, high, low, close) > 0
                and volume >= 0
                and high >= max(open_price, close, low)
                and low <= min(open_price, close, high)
            )
        except (TypeError, ValueError):
            valid = False
            open_time = 0
        if not valid:
            invalid += 1
            continue
        if open_time in unique:
            duplicates += 1
        normalized = dict(row)
        normalized["time"] = open_time
        normalized["marketEventId"] = canonical_market_event_id(symbol, tf, open_time)
        normalized["candleIsComplete"] = open_time + interval_ms <= now_ms
        unique[open_time] = normalized

    candles = [unique[key] for key in sorted(unique)]
    missing = 0
    for left, right in zip(candles, candles[1:]):
        delta = int(right["time"]) - int(left["time"])
        if delta > interval_ms:
            missing += max(0, round(delta / interval_ms) - 1)
    completed = [row for row in candles if row["candleIsComplete"]]
    incomplete = len(candles) - len(completed)
    freshness_ms = (
        max(0, now_ms - (int(completed[-1]["time"]) + interval_ms))
        if completed else math.inf
    )
    stale = not completed or freshness_ms > interval_ms * stale_after_intervals

    now = now_ms / 1000
    for kind, count in (
        ("DUPLICATE", duplicates),
        ("GAP", missing),
        ("OUT_OF_ORDER", out_of_order),
        ("INVALID_VALUE", invalid),
        ("INCOMPLETE", incomplete),
        ("STALE", int(stale)),
    ):
        if count:
            _incident(kind, symbol, tf, f"count={count}", now)
    _metrics["candle_batches"] += 1
    _metrics["accepted_candles"] += len(completed)
    return {
        "candles": candles,
        "completed": completed,
        "duplicates": duplicates,
        "missing": missing,
        "outOfOrder": out_of_order,
        "invalid": invalid,
        "incomplete": incomplete,
        "freshnessMs": freshness_ms,
        "stale": stale,
    }


@dataclass
class SnapshotDecision:
    accepted: bool
    duplicate: bool = False
    out_of_order: bool = False
    gap_seconds: float = 0.0
    freshness_seconds: float = 0.0
    recovered: bool = False


class SnapshotTracker:
    def __init__(self, expected_interval_seconds: float = 15.0):
        self.expected_interval_seconds = expected_interval_seconds
        self.last_timestamp: dict[str, float] = {}
        self.gapped: set[str] = set()

    def accept(self, symbol: str, source_timestamp: float, now: float | None = None) -> SnapshotDecision:
        now = time.time() if now is None else float(now)
        sym = normalize_symbol(symbol)
        ts = float(source_timestamp)
        if not math.isfinite(ts) or ts <= 0 or ts > now + 2:
            _incident("TIMESTAMP_VIOLATION", sym, "snapshot", f"source_timestamp={ts}", now)
            return SnapshotDecision(False)
        previous = self.last_timestamp.get(sym)
        if previous is not None and ts == previous:
            _incident("DUPLICATE", sym, "snapshot", f"timestamp={ts}", now)
            return SnapshotDecision(False, duplicate=True)
        if previous is not None and ts < previous:
            _incident("OUT_OF_ORDER", sym, "snapshot", f"{ts} < {previous}", now)
            return SnapshotDecision(False, out_of_order=True)
        gap = max(0.0, ts - previous) if previous is not None else 0.0
        is_gap = gap > self.expected_interval_seconds * 2
        recovered = sym in self.gapped and not is_gap
        if is_gap:
            self.gapped.add(sym)
            _incident("GAP", sym, "snapshot", f"gap_seconds={gap:.3f}", now)
        elif recovered:
            self.gapped.remove(sym)
            _incident("RECOVERED", sym, "snapshot", "continuity restored", now)
        self.last_timestamp[sym] = ts
        freshness = max(0.0, now - ts)
        if freshness > self.expected_interval_seconds * 2:
            _incident("STALE", sym, "snapshot", f"freshness_seconds={freshness:.3f}", now)
        _metrics["accepted_snapshots"] += 1
        return SnapshotDecision(True, gap_seconds=gap, freshness_seconds=freshness, recovered=recovered)


def validate_feature_contract(payload: dict[str, Any], now_ms: int | None = None) -> list[str]:
    now_ms = int(time.time() * 1000) if now_ms is None else int(now_ms)
    rejects: list[str] = []
    event_id = payload.get("marketEventId")
    feature_ts = payload.get("featureTimestampMs")
    if event_id and not str(event_id).startswith("md:v1:bingx:"):
        rejects.append("MARKET_EVENT_ID_REJECT: non-canonical market event id")
    if payload.get("candleIsComplete") is False:
        rejects.append("CANDLE_INCOMPLETE_REJECT: feature candle is unfinished")
    if feature_ts is not None:
        try:
            timestamp = int(feature_ts)
            if timestamp > now_ms + 2_000:
                rejects.append("FEATURE_TIMESTAMP_REJECT: feature timestamp is in the future")
            if now_ms - timestamp > 10 * 60_000:
                rejects.append("FEATURE_STALE_REJECT: feature timestamp is stale")
        except (TypeError, ValueError):
            rejects.append("FEATURE_TIMESTAMP_REJECT: feature timestamp is invalid")
    elif event_id:
        rejects.append("FEATURE_TIMESTAMP_REJECT: feature timestamp is missing")
    return rejects


def claim_event(event_id: str, side: str, now: float | None = None, ttl_seconds: float = 86400) -> bool:
    now = time.time() if now is None else float(now)
    for key, claimed_at in list(_event_claims.items()):
        if now - claimed_at > ttl_seconds:
            del _event_claims[key]
    key = f"{event_id}|{side}"
    if key in _event_claims:
        _incident("DUPLICATE_EXECUTION", event_id, "edge", side, now)
        return False
    _event_claims[key] = now
    return True


def quality_status() -> dict[str, Any]:
    return {
        "metrics": dict(_metrics),
        "incidents": list(_incidents),
        "activeEventClaims": len(_event_claims),
    }


def reset_quality_state() -> None:
    _metrics.clear()
    _incidents.clear()
    _event_claims.clear()

File: backend/src/lib/marketDataQuality.ts
export type MarketDataIncidentType =
  | "DUPLICATE"
  | "GAP"
  | "STALE"
  | "INCOMPLETE"
  | "OUT_OF_ORDER"
  | "INVALID_VALUE"
  | "TIMESTAMP_VIOLATION"
  | "DUPLICATE_EXECUTION";

export interface CandleInput {
  openTime: number;
  open: number;
  high: number;
  low: number;
  close: number;
  volume: number;
}

export interface MarketDataIncident {
  type: MarketDataIncidentType;
  symbol: string;
  timeframe: string;
  occurredAt: number;
  detail: string;
}

export interface CandleBatchQuality {
  candles: CandleInput[];
  completedCandles: CandleInput[];
  latestCompleted?: CandleInput;
  duplicateCount: number;
  missingCount: number;
  outOfOrderCount: number;
  invalidCount: number;
  incompleteCount: number;
  stale: boolean;
  freshnessMs: number;
  incidents: MarketDataIncident[];
}

const TIMEFRAME_MS: Record<string, number> = {
  "1m": 60_000,
  "3m": 180_000,
  "5m": 300_000,
  "15m": 900_000,
  "30m": 1_800_000,
  "1h": 3_600_000,
  "4h": 14_400_000,
  "1d": 86_400_000,
};

const metrics = {
  batches: 0,
  acceptedCandles: 0,
  duplicates: 0,
  missing: 0,
  outOfOrder: 0,
  invalid: 0,
  incomplete: 0,
  stale: 0,
  duplicateExecutions: 0,
};
const recentIncidents: MarketDataIncident[] = [];
const executionClaims = new Map<string, number>();

export function normalizeMarketSymbol(symbol: string): string {
  const compact = symbol.trim().toUpperCase().replace(/[_/]/g, "-");
  if (compact.includes("-")) return compact;
  for (const quote of ["USDT", "USDC", "USD"]) {
    if (compact.endsWith(quote)) {
      const base = compact.slice(0, -quote.length);
      return `${base}-${quote === "USD" ? "USDT" : quote}`;
    }
  }
  return compact;
}

export function normalizeTimeframe(timeframe: string): string {
  const normalized = timeframe.trim().toLowerCase();
  if (!TIMEFRAME_MS[normalized]) throw new Error(`Unsupported timeframe: ${timeframe}`);
  return normalized;
}

export function timeframeToMs(timeframe: string): number {
  return TIMEFRAME_MS[normalizeTimeframe(timeframe)];
}

export function canonicalMarketEventId(
  symbol: string,
  timeframe: string,
  candleOpenTimeMs: number,
): string {
  const openTime = Math.trunc(candleOpenTimeMs);
  if (!Number.isSafeInteger(openTime) || openTime <= 0) {
    throw new Error(`Invalid candle open timestamp: ${candleOpenTimeMs}`);
  }
  return `md:v1:bingx:${normalizeMarketSymbol(symbol)}:${normalizeTimeframe(timeframe)}:${openTime}`;
}

function recordIncident(incident: MarketDataIncident): void {
  recentIncidents.push(incident);
  if (recentIncidents.length > 200) recentIncidents.shift();
}

function validCandle(candle: CandleInput): boolean {
  const values = [
    candle.openTime, candle.open, candle.high, candle.low, candle.close, candle.volume,
  ];
  return values.every(Number.isFinite)
    && Number.isSafeInteger(candle.openTime)
    && candle.openTime > 0
    && candle.open > 0
    && candle.high > 0
    && candle.low > 0
    && candle.close > 0
    && candle.volume >= 0
    && candle.high >= Math.max(candle.open, candle.close, candle.low)
    && candle.low <= Math.min(candle.open, candle.close, candle.high);
}

export function assessCandleBatch(
  symbol: string,
  timeframe: string,
  input: CandleInput[],
  nowMs: number = Date.now(),
  staleAfterIntervals: number = 2,
): CandleBatchQuality {
  const normalizedSymbol = normalizeMarketSymbol(symbol);
  const normalizedTimeframe = normalizeTimeframe(timeframe);
  const intervalMs = timeframeToMs(normalizedTimeframe);
  const incidents: MarketDataIncident[] = [];
  let outOfOrderCount = 0;
  for (let i = 1; i < input.length; i++) {
    if (input[i].openTime < input[i - 1].openTime) outOfOrderCount++;
  }

  const byOpenTime = new Map<number, CandleInput>();
  let duplicateCount = 0;
  let invalidCount = 0;
  for (const candle of input) {
    if (!validCandle(candle) || candle.openTime % intervalMs !== 0) {
      invalidCount++;
      continue;
    }
    if (byOpenTime.has(candle.openTime)) duplicateCount++;
    byOpenTime.set(candle.openTime, candle);
  }
  const candles = [...byOpenTime.values()].sort((a, b) => a.openTime - b.openTime);

  let missingCount = 0;
  for (let i = 1; i < candles.length; i++) {
    const delta = candles[i].openTime - candles[i - 1].openTime;
    if (delta > intervalMs) missingCount += Math.max(0, Math.round(delta / intervalMs) - 1);
  }

  const completedCandles = candles.filter((c) => c.openTime + intervalMs <= nowMs);
  const incompleteCount = candles.length - completedCandles.length;
  const latestCompleted = completedCandles.at(-1);
  const freshnessMs = latestCompleted
    ? Math.max(0, nowMs - (latestCompleted.openTime + intervalMs))
    : Number.POSITIVE_INFINITY;
  const stale = !latestCompleted || freshnessMs > intervalMs * staleAfterIntervals;

  const add = (type: MarketDataIncidentType, count: number, detail: string) => {
    if (count <= 0) return;
    const incident = {
      type, symbol: normalizedSymbol, timeframe: normalizedTimeframe, occurredAt: nowMs, detail,
    };
    incidents.push(incident);
    recordIncident(incident);
  };
  add("DUPLICATE", duplicateCount, `${duplicateCount} duplicate candle row(s)`);
  add("GAP", missingCount, `${missingCount} missing candle interval(s)`);
  add("OUT_OF_ORDER", outOfOrderCount, `${outOfOrderCount} out-of-order transition(s)`);
  add("INVALID_VALUE", invalidCount, `${invalidCount} invalid candle row(s)`);
  add("INCOMPLETE", incompleteCount, `${incompleteCount} unfinished candle row(s) excluded`);
  add("STALE", stale ? 1 : 0, `latest completed candle freshness=${freshnessMs}ms`);

  metrics.batches++;
  metrics.acceptedCandles += completedCandles.length;
  metrics.duplicates += duplicateCount;
  metrics.missing += missingCount;
  metrics.outOfOrder += outOfOrderCount;
  metrics.invalid += invalidCount;
  metrics.incomplete += incompleteCount;
  if (stale) metrics.stale++;

  return {
    candles,
    completedCandles,
    latestCompleted,
    duplicateCount,
    missingCount,
    outOfOrderCount,
    invalidCount,
    incompleteCount,
    stale,
    freshnessMs,
    incidents,
  };
}

export function claimMarketEventExecution(
  marketEventId: string,
  positionSide: "LONG" | "SHORT",
  nowMs: number = Date.now(),
  retentionMs: number = 86_400_000,
): boolean {
  for (const [key, claimedAt] of executionClaims) {
    if (nowMs - claimedAt > retentionMs) executionClaims.delete(key);
  }
  const key = `${marketEventId}|${positionSide}`;
  if (executionClaims.has(key)) {
    metrics.duplicateExecutions++;
    recordIncident({
      type: "DUPLICATE_EXECUTION",
      symbol: marketEventId,
      timeframe: "execution",
      occurredAt: nowMs,
      detail: `execution already claimed for ${positionSide}`,
    });
    return false;
  }
  executionClaims.set(key, nowMs);
  return true;
}

export function releaseMarketEventExecution(
  marketEventId: string,
  positionSide: "LONG" | "SHORT",
): void {
  executionClaims.delete(`${marketEventId}|${positionSide}`);
}

export function getMarketDataQualityStatus() {
  return {
    metrics: { ...metrics },
    incidents: [...recentIncidents],
    activeExecutionClaims: executionClaims.size,
  };
}

export function resetMarketDataQualityState(): void {
  for (const key of Object.keys(metrics) as Array<keyof typeof metrics>) metrics[key] = 0;
  recentIncidents.length = 0;
  executionClaims.clear();
}
